src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,274 @@
+            {
+                _collectTask.Wait();
+                _collectTask = null;
+            }
+            _listener.Dispose();
+        }
+
+        private void RemoveInstrumentState(Instrument instrument, InstrumentState state)
+        {
+            _instrumentStates.TryRemove(instrument, out InstrumentState? _state);
+        }
+
+        private InstrumentState? GetInstrumentState(Instrument instrument)
+        {
+            if (!_instrumentStates.TryGetValue(instrument, out InstrumentState? instrumentState))
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
+                    {
+                        if (filter(instrument))
+                        {
+                            instrumentState = BuildInstrumentState(instrument);
+                            break;
+                        }
+                    }
+                }
+                if (instrumentState != null)
+                {
+                    _instrumentStates.TryAdd(instrument, instrumentState);
+                    instrumentState = _instrumentStates[instrument];
+                }
+            }
+            return instrumentState;
+        }
+
+        internal InstrumentState? BuildInstrumentState(Instrument instrument)
+        {
+            Func<Aggregator>? createAggregatorFunc = GetAggregatorFactory(instrument);
+            if (createAggregatorFunc == null)
+            {
+                return null;
+            }
+            Type aggregatorType = createAggregatorFunc.GetType().GenericTypeArguments[0];
+            Type instrumentStateType = typeof(InstrumentState<>).MakeGenericType(aggregatorType);
+            return (InstrumentState)Activator.CreateInstance(instrumentStateType, createAggregatorFunc)!;
+        }
+
+        private static Func<Aggregator>? GetAggregatorFactory(Instrument instrument)
+        {
+            Type type = instrument.GetType();
+            Type? genericDefType = null;
+#if IS_GENERIC_TYPE_SUPPORT
+            genericDefType = type.IsGenericType ? type.GetGenericTypeDefinition() : null;
+#else
+            try
+            {
+                 genericDefType = type.GetGenericTypeDefinition();
+            }
+            catch {}
+#endif
+            if (genericDefType == typeof(Counter<>))
+            {
+                return () => new RateSumAggregator();
+            }
+            else if (genericDefType == typeof(ObservableCounter<>))
+            {
+                return () => new RateAggregator();
+            }
+            else if (genericDefType == typeof(ObservableGauge<>))
+            {
+                return () => new LastValue();
+            }
+            else if (genericDefType == typeof(Histogram<>))
+            {
+                return () => new ExponentialHistogramAggregator(new PercentileAggregation(new double[] { 50, 95, 99 }));
```

Comment:
I assume the array is "constant", so move this to a static readonly field to avoid repeated allocations?